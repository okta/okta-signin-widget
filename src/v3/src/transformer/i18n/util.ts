/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
