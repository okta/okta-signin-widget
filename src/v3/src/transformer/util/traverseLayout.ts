/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
