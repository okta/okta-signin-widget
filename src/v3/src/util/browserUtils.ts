/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const isAndroid = (): boolean => (
  // Windows Phone also contains "Android"
  /android/i.test(navigator.userAgent)
    && !/windows phone/i.test(navigator.userAgent)
);

export const isIOS = (): boolean => (
  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  /iPad|iPhone|iPod/i.test(navigator.userAgent)
    // iPad on iOS 13 detection
    || (navigator.userAgent.includes('Mac') && 'ontouchend' in document)
);

export const isAndroidOrIOS = (): boolean => isAndroid() || isIOS();
