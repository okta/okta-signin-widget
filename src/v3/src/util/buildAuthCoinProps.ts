/*
 * Copyright (c) 2025-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  AUTHENTICATOR_KEY, DEVICE_ENROLLMENT_TYPE, EMAIL_AUTHENTICATOR_TERMINAL_KEYS, IDX_STEP,
} from '../constants';
import { AuthCoinProps } from '../types';
import { getAuthenticatorKey } from './getAuthenticatorKey';
import { containsOneOfMessageKeys } from './idx';

export const buildAuthCoinProps = (
  transaction?: IdxTransaction | null,
): AuthCoinProps | undefined => {
  if (!transaction) {
    return undefined;
  }

  const { nextStep, messages } = transaction;
  if (containsOneOfMessageKeys(EMAIL_AUTHENTICATOR_TERMINAL_KEYS, messages)
    || nextStep?.name === IDX_STEP.CONSENT_EMAIL_CHALLENGE) {
    return { authenticatorKey: AUTHENTICATOR_KEY.EMAIL };
  }

  if (nextStep?.name === IDX_STEP.PIV_IDP) {
    return { authenticatorKey: IDX_STEP.PIV_IDP };
  }

  if (nextStep?.name === IDX_STEP.REDIRECT_IDVERIFY && nextStep?.idp?.id) {
    return { authenticatorKey: nextStep.idp.id };
  }

  // @ts-expect-error Property 'deviceEnrollment' does not exist on type 'IdxContext' ts(2339)
  if (transaction.context?.deviceEnrollment?.value?.name === DEVICE_ENROLLMENT_TYPE.ODA) {
    return { authenticatorKey: AUTHENTICATOR_KEY.OV };
  }

  const authenticatorKey = getAuthenticatorKey(transaction);
  if (authenticatorKey) {
    return {
      authenticatorKey,
      // @ts-ignore logoUri missing from interface
      url: nextStep?.relatesTo?.value.logoUri,
    };
  }

  return undefined;
};
