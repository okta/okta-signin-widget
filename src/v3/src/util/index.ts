/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
