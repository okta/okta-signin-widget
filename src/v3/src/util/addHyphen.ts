/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
*/
import { KeyboardEvent } from "src/types"

export const addHyphen = (e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    const currentVal = e.currentTarget.value;
    // Add hyphen after 4th character. includes() check allows deletion of hyphen and prevents double hyphens
    if (currentVal && currentVal.length === 4 && !['Backspace', 'Delete', '-'].includes(e.key)) {
      e.currentTarget.value = currentVal.concat('-');
    } else if (currentVal && currentVal.length > 4 && currentVal.charAt(4) !== '-') {
      // If user types too fast, the above concatenation can be skipped. This retroactively adds a hyphen if it wasn't added
      e.currentTarget.value = currentVal.slice(0, 4) + '-' + currentVal.slice(4);
    }
}
