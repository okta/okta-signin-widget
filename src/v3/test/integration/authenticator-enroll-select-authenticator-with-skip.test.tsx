/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { createAuthJsPayloadArgs, setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/authenticator-enroll-select-authenticator-with-skip.json';

describe('authenticator-enroll-select-authenticator', () => {
  it('renders form', async () => {
    const { container, findByText } = await setup({ mockResponse });
    expect(await findByText(/Set up security methods/)).toBeInTheDocument();
    expect(await findByText(/Set up optional/)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  describe('send correct payload', () => {
    it('when select Phone', async () => {
      const {
        authClient,
        user,
        findByTestId,
      } = await setup({ mockResponse });

      const authenticatorButton = await findByTestId('phone_number');
      await user.click(authenticatorButton);
      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/credential/enroll', {
          authenticator: { id: 'aid568g3mXgtID0X1SLH' },
        }),
      );
    });

    it('skips the step when clicking continue', async () => {
      const {
        authClient,
        user,
        findByText,
      } = await setup({ mockResponse });

      const skipBtn = await findByText('Continue', { selector: 'button' });
      await user.click(skipBtn);
      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/skip'),
      );
    });
  });
});
