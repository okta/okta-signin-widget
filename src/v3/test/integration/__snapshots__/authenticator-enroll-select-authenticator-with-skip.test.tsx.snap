// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticator-enroll-select-authenticator renders form 1`] = `
<div>
  <main
    class="auth-container main-container MuiBox-root emotion-0"
    data-commit="b9bbc0140703c3fbf0e2e58920362e70"
    data-version="0.0.0"
    dir="ltr"
    id="okta-sign-in"
    lang="en"
  >
    <div
      class="MuiScopedCssBaseline-root emotion-1"
    >
      <div
        class="MuiBox-root emotion-2"
      >
        <div
          class="MuiBox-root emotion-3"
        >
          <div
            class="okta-sign-in-header auth-header MuiBox-root emotion-4"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 emotion-5"
            />
          </div>
          <div
            class="MuiBox-root emotion-6"
          >
            <div
              class="identifier-container MuiBox-root emotion-7"
            >
              <div
                class="MuiBox-root emotion-8"
              >
                <span
                  class="MuiBox-root emotion-9"
                >
                  <svg
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-10"
                    fill="none"
                    focusable="false"
                    role="img"
                    viewBox="0 0 16 16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M10 3V5C10 6.10457 9.10457 7 8 7C6.89543 7 6 6.10457 6 5V3C6 1.89543 6.89543 1 8 1C9.10457 1 10 1.89543 10 3ZM5 3C5 1.34315 6.34315 0 8 0C9.65685 0 11 1.34315 11 3V5C11 6.65685 9.65685 8 8 8C6.34315 8 5 6.65685 5 5V3ZM4.1305 10.0742C4.26803 10.0181 4.459 10 5.41376 10H10.5862C11.541 10 11.732 10.0181 11.8695 10.0742C12.0299 10.1398 12.1706 10.2459 12.2777 10.3821C12.3695 10.4989 12.4393 10.6776 12.7016 11.5956L13.6743 15H2.32573L3.29841 11.5956C3.5607 10.6776 3.63054 10.4989 3.72233 10.3821C3.82941 10.2459 3.97006 10.1398 4.1305 10.0742ZM13.6631 11.3209L14.7143 15L15 16H13.96H2.04002H1L1.28571 15L2.33689 11.3209C2.57458 10.489 2.69343 10.073 2.93606 9.76423C3.15022 9.49171 3.43152 9.27953 3.75239 9.14848C4.11592 9 4.54854 9 5.41376 9H10.5862C11.4515 9 11.8841 9 12.2476 9.14848C12.5685 9.27953 12.8498 9.49171 13.0639 9.76423C13.3066 10.073 13.4254 10.489 13.6631 11.3209Z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <title>
                      User
                    </title>
                  </svg>
                </span>
                <span
                  class="identifier no-translate MuiBox-root emotion-11"
                  data-se="identifier"
                  title="testUser@okta.com"
                >
                  testUser@okta.com
                </span>
              </div>
            </div>
            <form
              class="o-form MuiBox-root emotion-12"
              data-se="o-form"
              novalidate=""
            >
              <div
                class="MuiBox-root emotion-13"
              >
                <div
                  class="MuiBox-root emotion-14"
                >
                  <div
                    class="MuiBox-root emotion-15"
                  >
                    <h2
                      class="MuiTypography-root MuiTypography-h4 emotion-16"
                      data-se="o-form-head"
                      id="select-authenticator-enroll_Title_Set_up_security_methods_1"
                      tabindex="-1"
                    >
                      Set up security methods
                    </h2>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-14"
                >
                  <div
                    class="MuiBox-root emotion-15"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-19"
                      data-se="o-form-explain"
                      id="select-authenticator-enroll_Description_Security_methods_help_protect_your_account_by_ensuring_only_you_have_access_2"
                    >
                      Security methods help protect your account by ensuring only you have access.
                    </p>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-0"
                >
                  <div
                    class="MuiBox-root emotion-15"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-19"
                      data-se="o-form-explain"
                      id="select-authenticator-enroll_Description_Set_up_optional_3"
                    >
                      Set up optional
                    </p>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-14"
                >
                  <ul
                    class="MuiBox-root emotion-24"
                    data-se="authenticator-enroll-list"
                  >
                    <li
                      class="MuiBox-root emotion-14"
                    >
                      <button
                        aria-describedby="okta_password_0-description okta_password_0-ctaLabel"
                        aria-labelledby="okta_password_0-label"
                        class="authenticator-row MuiBox-root emotion-26"
                        data-se="authenticator-button"
                        tabindex="0"
                        type="button"
                      >
                        <div
                          class="authenticator-icon-container MuiBox-root emotion-0"
                          data-se="authenticator-icon"
                        >
                          <div
                            aria-hidden="true"
                            class="mfa-okta-password authenticator-icon MuiBox-root emotion-28"
                            data-se="factor-beacon"
                          >
                            <svg
                              aria-labelledby="okta_password_0"
                              fill="none"
                              height="48"
                              role="img"
                              viewBox="0 0 48 48"
                              width="48"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <title
                                id="okta_password_0"
                              >
                                Password
                              </title>
                              <circle
                                class="siwFillBg"
                                cx="24"
                                cy="24"
                                fill="#F5F5F6"
                                r="24"
                              />
                              <path
                                class="siwFillSecondary"
                                d="M11 17h-1V9h8v1h-7v7Zm28 0h-1v-7h-7V9h8v8ZM18 39h-8v-8h1v7h7v1Zm21 0h-8v-1h7v-7h1v8Z"
                                fill="#A7B5EC"
                              />
                              <path
                                class="siwFillPrimaryDark"
                                d="m14.91 22.067-1.829 1.064V21h-1v2.131l-1.83-1.064-.502.866L11.585 24l-1.836 1.067.502.866 1.83-1.064V27h1v-2.131l1.829 1.064.503-.866L13.576 24l1.837-1.067-.503-.866Zm7.946 0-1.829 1.064V21h-1v2.131l-1.829-1.064-.503.866L19.531 24l-1.836 1.067.503.866 1.83-1.064V27h1v-2.131l1.828 1.064.503-.866L21.523 24l1.836-1.067-.503-.866Zm7.946 0-1.829 1.064V21h-1v2.131l-1.829-1.064-.503.866L27.477 24l-1.836 1.067.503.866 1.83-1.064V27h1v-2.131l1.828 1.064.503-.866L29.47 24l1.836-1.067-.503-.866Zm8.448.866-.501-.866-1.83 1.064V21h-1v2.131l-1.83-1.064-.502.866L35.424 24l-1.837 1.067.503.866 1.829-1.064V27h1v-2.131l1.83 1.064.502-.866L37.415 24l1.836-1.067Z"
                                fill="#00297A"
                              />
                            </svg>
                          </div>
                        </div>
                        <div
                          class="authenticator-description MuiBox-root emotion-29"
                        >
                          <h3
                            class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-30"
                            data-se="authenticator-button-label"
                            id="okta_password_0-label"
                          >
                            Password
                          </h3>
                          <p
                            class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph authenticator-description--text emotion-31"
                            data-se="authenticator-button-description"
                            id="okta_password_0-description"
                          >
                            Choose a password for your account
                          </p>
                          <div
                            class="cta-button authenticator-button MuiBox-root emotion-32"
                            data-se="okta_password"
                          >
                            <span
                              class="button select-factor link-button MuiBox-root emotion-33"
                              data-se="cta-button-label"
                              id="okta_password_0-ctaLabel"
                            >
                              Set up
                            </span>
                            <svg
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-34"
                              fill="none"
                              focusable="false"
                              role="img"
                              viewBox="0 0 16 16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                clip-rule="evenodd"
                                d="M13.2929 7.5L9.14645 3.35355L9.85355 2.64645L14.8536 7.64645C15.0488 7.84171 15.0488 8.15829 14.8536 8.35355L9.85355 13.3536L9.14645 12.6464L13.2929 8.5H1V7.5H13.2929Z"
                                fill="currentColor"
                                fill-rule="evenodd"
                              />
                              <title>
                                Set up
                              </title>
                            </svg>
                          </div>
                        </div>
                      </button>
                    </li>
                    <li
                      class="MuiBox-root emotion-14"
                    >
                      <button
                        aria-describedby="phone_number_1-description phone_number_1-ctaLabel"
                        aria-labelledby="phone_number_1-label"
                        class="authenticator-row MuiBox-root emotion-26"
                        data-se="authenticator-button"
                        tabindex="0"
                        type="button"
                      >
                        <div
                          class="authenticator-icon-container MuiBox-root emotion-0"
                          data-se="authenticator-icon"
                        >
                          <div
                            aria-hidden="true"
                            class="mfa-okta-phone authenticator-icon MuiBox-root emotion-28"
                            data-se="factor-beacon"
                          >
                            <svg
                              aria-labelledby="phone_number_1"
                              fill="none"
                              height="48"
                              role="img"
                              viewBox="0 0 48 48"
                              width="48"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <title
                                id="phone_number_1"
                              >
                                Voice Call Authentication
                              </title>
                              <circle
                                class="siwFillBg"
                                cx="24"
                                cy="24"
                                fill="#F5F5F6"
                                r="24"
                              />
                              <path
                                class="siwFillPrimaryDark"
                                d="M21.5 37h-9a2.6 2.6 0 0 1-2.5-2.682V14.682A2.6 2.6 0 0 1 12.5 12h9a2.6 2.6 0 0 1 2.5 2.682v19.636A2.6 2.6 0 0 1 21.5 37Zm-9-24a1.6 1.6 0 0 0-1.5 1.682v19.636A1.6 1.6 0 0 0 12.5 36h9a1.6 1.6 0 0 0 1.5-1.682V14.682A1.6 1.6 0 0 0 21.5 13h-9Z"
                                fill="#00297A"
                              />
                              <path
                                class="siwFillPrimaryDark"
                                d="M19 34h-4v1h4v-1Zm-1-19a1 1 0 1 0-2 0 1 1 0 0 0 2 0Z"
                                fill="#00297A"
                              />
                              <path
                                class="siwFillSecondary"
                                d="M34.606 34.606 33.9 33.9a13.998 13.998 0 0 0 0-19.8l.707-.707a15 15 0 0 1 0 21.212l-.001.001Z"
                                fill="#A7B5EC"
                              />
                              <path
                                class="siwFillSecondary"
                                d="m31.071 31.071-.707-.707a9 9 0 0 0 0-12.728l.707-.707a10 10 0 0 1 0 14.142Zm-3.535-3.536-.707-.707a4 4 0 0 0 0-5.656l.707-.707a5 5 0 0 1 0 7.07Z"
                                fill="#A7B5EC"
                              />
                            </svg>
                          </div>
                        </div>
                        <div
                          class="authenticator-description MuiBox-root emotion-29"
                        >
                          <h3
                            class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-30"
                            data-se="authenticator-button-label"
                            id="phone_number_1-label"
                          >
                            Phone
                          </h3>
                          <p
                            class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph authenticator-description--text emotion-31"
                            data-se="authenticator-button-description"
                            id="phone_number_1-description"
                          >
                            Verify with a code sent to your phone
                          </p>
                          <div
                            class="cta-button authenticator-button MuiBox-root emotion-32"
                            data-se="phone_number"
                          >
                            <span
                              class="button select-factor link-button MuiBox-root emotion-33"
                              data-se="cta-button-label"
                              id="phone_number_1-ctaLabel"
                            >
                              Set up
                            </span>
                            <svg
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-34"
                              fill="none"
                              focusable="false"
                              role="img"
                              viewBox="0 0 16 16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                clip-rule="evenodd"
                                d="M13.2929 7.5L9.14645 3.35355L9.85355 2.64645L14.8536 7.64645C15.0488 7.84171 15.0488 8.15829 14.8536 8.35355L9.85355 13.3536L9.14645 12.6464L13.2929 8.5H1V7.5H13.2929Z"
                                fill="currentColor"
                                fill-rule="evenodd"
                              />
                              <title>
                                Set up
                              </title>
                            </svg>
                          </div>
                        </div>
                      </button>
                    </li>
                    <li
                      class="MuiBox-root emotion-14"
                    >
                      <button
                        aria-describedby="webauthn_2-description webauthn_2-ctaLabel"
                        aria-labelledby="webauthn_2-label"
                        class="authenticator-row MuiBox-root emotion-26"
                        data-se="authenticator-button"
                        tabindex="0"
                        type="button"
                      >
                        <div
                          class="authenticator-icon-container MuiBox-root emotion-0"
                          data-se="authenticator-icon"
                        >
                          <div
                            aria-hidden="true"
                            class="mfa-webauthn authenticator-icon MuiBox-root emotion-28"
                            data-se="factor-beacon"
                          >
                            <svg
                              aria-labelledby="webauthn_2"
                              fill="none"
                              height="48"
                              role="img"
                              viewBox="0 0 48 48"
                              width="48"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <title
                                id="webauthn_2"
                              >
                                Security Key or Biometric Authenticator
                              </title>
                              <path
                                class="siwFillBg"
                                clip-rule="evenodd"
                                d="M24 48c13.255 0 24-10.745 24-24S37.255 0 24 0 0 10.745 0 24s10.745 24 24 24Z"
                                fill="#F5F5F6"
                                fill-rule="evenodd"
                              />
                              <path
                                class="siwFillPrimaryDark"
                                d="M24 17V9h-9v8h-3v15.5a2.5 2.5 0 0 0 2.5 2.5h6.34a8.58 8.58 0 0 1-.38-1H14.5a1.5 1.5 0 0 1-1.5-1.5V18h13v3.84a8.58 8.58 0 0 1 1-.38V17h-3Zm-8 0v-7h7v7h-7Z"
                                fill="#00297A"
                              />
                              <path
                                class="siwFillPrimaryDark"
                                d="M19.5 25a2.5 2.5 0 1 1 0-5 2.5 2.5 0 0 1 0 5Zm0-4a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3Z"
                                fill="#00297A"
                              />
                              <path
                                class="siwFillSecondary"
                                d="M30 40a9 9 0 1 1 9-9 9.01 9.01 0 0 1-9 9Zm0-17a8 8 0 1 0 8 8 8.009 8.009 0 0 0-8-8Z"
                                fill="#A7B5EC"
                              />
                              <path
                                class="siwFillSecondary"
                                d="M32.084 36.019a.472.472 0 0 1-.319-.112.501.501 0 0 1-.159-.532c.43-1.389 1.274-4.787-.119-6.352a1.962 1.962 0 0 0-1.946-.6 1.93 1.93 0 0 0-1.387 2.354c.007.024.666 2.823-.966 4.55a.5.5 0 1 1-.727-.685c1.24-1.314.728-3.6.723-3.622a2.933 2.933 0 0 1 2.106-3.566 2.977 2.977 0 0 1 2.944.906c1.67 1.875.922 5.386.328 7.312a.5.5 0 0 1-.478.347Z"
                                fill="#A7B5EC"
                              />
                              <path
                                class="siwFillSecondary"
                                d="M28.63 37.25a.5.5 0 0 1-.359-.849c2.235-2.3 1.345-5.782 1.336-5.817a.5.5 0 0 1 .966-.259c.045.165 1.052 4.067-1.586 6.774a.5.5 0 0 1-.358.151Z"
                                fill="#A7B5EC"
                              />
                              <path
                                class="siwFillSecondary"
                                d="M34.88 35.2a.5.5 0 0 1-.488-.612c.493-2.2.716-5.235-1.03-7.2a4.508 4.508 0 0 0-4.455-1.357 4.416 4.416 0 0 0-2.72 2.069 4.2 4.2 0 0 0-.468 3.2 3.1 3.1 0 0 1-.355 2.347.5.5 0 1 1-.728-.685 2.37 2.37 0 0 0 .11-1.434 5.174 5.174 0 0 1 .577-3.939 5.42 5.42 0 0 1 3.333-2.518 5.524 5.524 0 0 1 5.453 1.66c1.628 1.832 2.051 4.551 1.257 8.082a.5.5 0 0 1-.487.387Z"
                                fill="#A7B5EC"
                              />
                            </svg>
                          </div>
                        </div>
                        <div
                          class="authenticator-description MuiBox-root emotion-29"
                        >
                          <h3
                            class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-30"
                            data-se="authenticator-button-label"
                            id="webauthn_2-label"
                          >
                            Security Key or Biometric Authenticator
                          </h3>
                          <p
                            class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph authenticator-description--text emotion-31"
                            data-se="authenticator-button-description"
                            id="webauthn_2-description"
                          >
                            Use a security key or a biometric authenticator to sign in
                          </p>
                          <div
                            class="cta-button authenticator-button MuiBox-root emotion-32"
                            data-se="webauthn"
                          >
                            <span
                              class="button select-factor link-button MuiBox-root emotion-33"
                              data-se="cta-button-label"
                              id="webauthn_2-ctaLabel"
                            >
                              Set up
                            </span>
                            <svg
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-34"
                              fill="none"
                              focusable="false"
                              role="img"
                              viewBox="0 0 16 16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                clip-rule="evenodd"
                                d="M13.2929 7.5L9.14645 3.35355L9.85355 2.64645L14.8536 7.64645C15.0488 7.84171 15.0488 8.15829 14.8536 8.35355L9.85355 13.3536L9.14645 12.6464L13.2929 8.5H1V7.5H13.2929Z"
                                fill="currentColor"
                                fill-rule="evenodd"
                              />
                              <title>
                                Set up
                              </title>
                            </svg>
                          </div>
                        </div>
                      </button>
                    </li>
                  </ul>
                </div>
                <div
                  class="MuiBox-root emotion-14"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth emotion-56"
                    data-type="save"
                    tabindex="0"
                    type="submit"
                  >
                    Set up later
                  </button>
                </div>
                <div
                  class="MuiBox-root emotion-14"
                >
                  <button
                    class="MuiTypography-root MuiTypography-body1 MuiLink-root MuiLink-underlineAlways MuiLink-button emotion-58"
                    data-se="cancel"
                    role="link"
                  >
                    Back to sign in
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
